### PR TITLE
[sanitizer_common] Handle ptrace on Linux/sparc64

### PR DESCRIPTION
**Pull request by rorth**
When ASan testing is enabled on SPARC as per PR #107405, the
```
  AddressSanitizer-sparc-linux :: TestCases/Linux/ptrace.cpp
```
`FAIL`s on Linux/sparc64.  This happens because the `ptrace` interceptor has no support for that target at all.

This patch adds the missing parts and accounts for a couple of issues specific to this target:
- In some cases, SPARC just needs to be included in the list of supported targets.
- Besides, the types used by the `PTRACE_GETREGS` and `PTRACE_GETFPREGS` requests need to be filled in.
- `ptrace` has a weird quirk on this target: for a couple of requests, the meaning of the `data` and `addr` args is reversed.  All of the `Linux/ptrace.cpp` test and the interceptor, pre-syscall and post-syscall hooks need to account for that swap in their checks.

Tested on `sparc64-unknown-linux-gnu` and `x86_64-pc-linux-gnu`.

### CODE REVIEW

File: compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc [hunk 1]
```diff
@@ -3430,23 +3430,34 @@ INTERCEPTOR(uptr, ptrace, int request, int pid, void *addr, void *data) {
   COMMON_INTERCEPTOR_ENTER(ctx, ptrace, request, pid, addr, data);
   __sanitizer_iovec local_iovec;
 
-  if (data) {
+  void *data_arg = data;
+#  if defined(__sparc__)
+  // The meanings of addr and data are reversed for a few requests on
+  // Linux/sparc64.
+  if (request == ptrace_getregs || request == ptrace_getfpregs ||
```
Comment (vitalybuka) [inline compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:3437]: That's really weird.
Can you add a comment with proof-link here for future readers?

File: compiler-rt/lib/sanitizer_common/sanitizer_common_syscalls.inc [hunk 2]
```diff
@@ -2530,18 +2530,26 @@ PRE_SYSCALL(ptrace)(long request, long pid, long addr, long data) {
 #  if !SANITIZER_ANDROID &&                                                   \
       (defined(__i386) || defined(__x86_64) || defined(__mips64) ||           \
        defined(__powerpc64__) || defined(__aarch64__) || defined(__s390__) || \
-       defined(__loongarch__) || SANITIZER_RISCV64)
-  if (data) {
+       defined(__loongarch__) || SANITIZER_RISCV64 || defined(__sparc__))
+  long data_arg = data;
```
Comment (vitalybuka) [inline compiler-rt/lib/sanitizer_common/sanitizer_common_syscalls.inc:2534]: maybe:

 compiler-rt/lib/sanitizer_common/sanitizer_common_ptrace.h

```
long data_arg = ptrace_data_or_arg(data, arg, request);

template<class T>
inline T ptrace_data_or_arg(T data, T arg, request) {
#    if defined(__sparc__)
  // The meanings of addr and data are reversed for a few requests on
  // Linux/sparc64.
  if (request == ptrace_getregs || request == ptrace_getfpregs ||
      request == ptrace_setregs || request == ptrace_setfpregs)
    return addr;
#else
  return data;
#    endif
}
```

**Review by vitalybuka — APPROVED** [review]
LGTM, but extracting helper would be nice

File: compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc (hunk 1 above)
Comment (rorth) [reply compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc]: Done: it's on `ptrace(2)`.

File: compiler-rt/lib/sanitizer_common/sanitizer_common_syscalls.inc (hunk 2 above)
Comment (rorth) [reply compiler-rt/lib/sanitizer_common/sanitizer_common_syscalls.inc]: That's really nice, thanks.  I made only a few minor adjustments, primarily having the `ptrace_data_arg` template match the `ptrace` arg order.

One caveat, though: it doesn't work to put this into `sanitizer_common_ptrace.h`: in that case, the `ptrace_*` declarations are missing.  I've chosen `sanitizer_platform_limits_posix.h` where the latter already are.  However, I had to include that file in both `sanitizer_common_interceptors.inc` ( a generic file, although the `ptrace` interceptor is in a Linux-only section) and `sanitizer_common_syscalls.inc` (Linux-specific, so shouldn't be an issue).

Please let me know if this is ok?

**Pull request merged**